net/tcp/tcp_conn.c

```diff
@@ -836,10 +863,25 @@ void tcp_free(FAR struct tcp_conn_s *conn)
     }
 #endif
 
-  /* Mark the connection available and put it into the free list */
+  /* Mark the connection available. */
 
   conn->tcpstateflags = TCP_CLOSED;
-  dq_addlast(&conn->sconn.node, &g_free_tcp_connections);
+
+  /* Check if this is a preallocated connection to store it in the free
+   * connections list, else deallocate it.
+   */
+
+#ifdef CONFIG_NET_ALLOC_CONNS
+  if ((conn->flags & TCP_PREALLOC) == 0)
+    {
+      kmm_free(conn);
```

Comment:
It doesn't matter. The preallocated connections are never free'd, and they are the only ones allocated in a batch.

The later connections, that are actually free'd, are also allocated one-by-one as needed.

Reply:
If so, you destroy NET_TCP_CONNS meaning:
```
config NET_TCP_CONNS
	int "Number of TCP/IP connections"
	default 8
	---help---
		Maximum number of TCP/IP connections (all tasks)
		Number of TCP/IP connections (all tasks).

		This number of connections will be pre-allocated during system boot.
		If NET_ALLOC_CONNS is enabled, more connections may be allocated at
		a later time, as the system needs them. Else this will be the 
		maximum number of connections avaiable to the system at all times.
```
We need keep the code consistent with config.

Reply:
Let's summary my suggestion:

1. NET_TCP_PREALLOC_CONNS to indicate the preallocated conn from .bss section

   - zero disable the prealloation

3. NET_TCP_ALLOC_CONNS to indicate the dynamic conn from heap:

   -  0 mean disable dynamical allocation at all
   - 1 mean allocate/free from heap every time
   - 2... mean allocate batch from heap, but not return to memory

4. Remove CONFIG_NET_ALLOC_CONNS once all conn convert to this style

Reply:
OK, we agree. (Especially on point 3!)

The only issue is that this way preallocated connections will be always in BSS.  
This means that changing this value will have the effect of *moving* the connection struct from BSS to heap (as it will be allocated if it is nevertheless needed).

This may be a bit problematic in arch'es with multiple memory regions.

It is not my favorite, but I can live with that.

---

One thing that I don't truly like is the `alloc_conn` variable as it is used now (tcp_conn.c, line 628).
I wanted to mark the connection as pre-allocated within `tcp_initialize()`, but the conn struct will be memset to 0 in line 735, so this information is lost.

Do you have any suggestions for this?

Reply:
> OK, we agree. (Especially on point 3!)
> 
> The only issue is that this way preallocated connections will be always in BSS. This means that changing this value will have the effect of _moving_ the connection struct from BSS to heap (as it will be allocated if it is nevertheless needed).
> 
> This may be a bit problematic in arch'es with multiple memory regions.
> 
> It is not my favorite, but I can live with that.
>

We can allocate the preallocated conn from heap too, it is no much difference whether they come from .bss or heap. The more important thing is never return them to heap.

> One thing that I don't truly like is the `alloc_conn` variable as it is used now (tcp_conn.c, line 628). I wanted to mark the connection as pre-allocated within `tcp_initialize()`, but the conn struct will be memset to 0 in line 735, so this information is lost.
> 
> Do you have any suggestions for this?

One simple approach is:

1. Don't preallocate conn in startup instead
2. Put conn to free list if free list is smaller than NET_TCP_PREALLOC_CONNS
3. Call kmm_free if the free list is larger than NET_TCP_PREALLOC_CONNS

So, we don't need flag to distinguish whether conn is preallocate or not.

Reply:
> One simple approach is:
> 
> 1. Don't preallocate conn in startup instead
> 2. Put conn to free list if free list is smaller than NET_TCP_PREALLOC_CONNS
> 3. Call kmm_free if the free list is larger than NET_TCP_PREALLOC_CONNS
> 
> So, we don't need flag to distinguish whether conn is preallocate or not.

This is what I thought first, but this is not working.  
We need to preserve the exact same connections allocated during init. If not, heap fragmentation will happen. The preallocated list will serve absolutely no purpose.

Reply:
Yes, it's a problem. So, I think the better approach is move flag mark and memset to tcp_alloc_conn.

Reply:
> Yes, it's a problem. So, I think the better approach is move flag mark and memset to tcp_alloc_conn.

I also thought of that, but now `memset()` is used for all connections (even those in the free list), every time (not only during the initial allocation).

I am not sure whether code anywhere else works under the assumption that conns are `memset()` at this point, or it is just defensive coding.

Also I am not sure whether
```g_tcp_connections[i].tcpstateflags = TCP_CLOSED;```
has any meaning in line 605, or it will be overwritten by the memset (and thus it is dead code).